### PR TITLE
Add GregorianDate class to represent SQL DATE type

### PR DESCRIPTION
**Pull request by samalone**
The new class maps to and from the "YYYY-MM-dd" string format, which is the default representation for the SQL DATE type in most databases. The type is also convertible
to the Swift Date and DateComponents types.

An upcoming pull request to vapor/fluent-sql-driver will add SQLite support.

### CODE REVIEW

**Comment by samalone** [comment]
I fixed the failing unit test on Linux by eliminating it. It was trying to decode a GregorianDate directly from a JSON string fragment, which works on Mac but fails on Linux.

File: Sources/FluentKit/Utilities/GregorianDate.swift [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+enum GregorianDateError: Error {
+    case invalidStringFormat
+}
+
+/**
+ A Gregorian date with no time or timezone, intended to pair with
+ the SQL `DATE` type
+ */
+public struct GregorianDate {
+    private static let gregorianCalendar = Calendar(identifier: .gregorian)
```
Comment (tanner0101) [inline Sources/FluentKit/Utilities/GregorianDate.swift]: Is `Calendar` thread safe?

File: Sources/FluentKit/Utilities/GregorianDate.swift [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    
+    public var year: Int
+    public var month: Int
+    public var day: Int
+    
+    public var dateComponents: DateComponents {
+        return DateComponents(calendar: GregorianDate.gregorianCalendar,
```
Comment (tanner0101) [inline Sources/FluentKit/Utilities/GregorianDate.swift]: No "Xcode" formatting, please. This should be like:
```swift
Foo(
    bar: ...,
    baz: ...
)
```

File: Sources/FluentKit/Utilities/GregorianDate.swift [hunk 3]
```diff
@@ -0,0 +1,142 @@
+import Foundation
+
+enum GregorianDateError: Error {
+    case invalidStringFormat
+}
+
+/**
```
Comment (tanner0101) [inline Sources/FluentKit/Utilities/GregorianDate.swift]: Doc blocks should be `///`

File: Sources/FluentKit/Utilities/GregorianDate.swift [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     Create a GregorianDate from a year, month, and day.
+     The inputs are not validated to ensure they form a valid date.
+     
+     - Parameters:
+         - year: >= 0
+         - month: 1 – 12
+         - day: 1 – 31
+     */
+    public init(year: Int, month: Int, day: Int) {
+        self.year = year
```
Comment (tanner0101) [inline Sources/FluentKit/Utilities/GregorianDate.swift]: Should we assert the docblock requirements here?

File: Sources/FluentKit/Utilities/GregorianDate.swift [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        self.day = day
+    }
+    
+    /**
+     Create a GregorianDate from the date portion of a Date object,
+     using the timezone stored in that Date.
+     */
+    public init(_ date: Date) {
+        let components = GregorianDate.gregorianCalendar.dateComponents([.year, .month, .day], from: date)
+        self.year = components.year!
```
Comment (tanner0101) [inline Sources/FluentKit/Utilities/GregorianDate.swift]: Can we ever hit assertion for these force unwraps? If so, this init could be nil fallible.

File: Sources/FluentKit/Utilities/GregorianDate.swift [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    }
+    
+    /**
+     Create a GregorianDate from a string in YYYY-MM-dd format.
+     */
+    public init(string: String) throws {
+        // Yes, this is a horrendous amount of code just to parse the "YYYY-MM-dd" date format.
+        // It could also be done with NSRegularExpression, or by converting to Date and using
+        // a DateFormatter. Feel free.
+        let zero = CChar(48)
```
Comment (tanner0101) [inline Sources/FluentKit/Utilities/GregorianDate.swift]: Could we do something like:
```swift
let parts = string.split("-")
guard parts.count == 3 else {
    return nil
}
let year = Int(parts[0])
let month = Int(parts[1])
let day = Int(parts[2])
```

File: Sources/FluentKit/Utilities/GregorianDate.swift [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        let components = GregorianDate.gregorianCalendar.dateComponents([.year, .month, .day], from: date)
+        self.year = components.year!
+        self.month = components.month!
+        self.day = components.day!
+    }
+    
+    /**
+     Create a GregorianDate from a string in YYYY-MM-dd format.
+     */
+    public init(string: String) throws {
```
Comment (tanner0101) [inline Sources/FluentKit/Utilities/GregorianDate.swift]: I'd make this nil fallible instead of throwing.

**Comment by tanner0101** [comment]
Closing for inactivity, feel free to re-open with updates, thanks.

**Pull request closed without merging**